power_grid_model_c/power_grid_model/include/power_grid_model/sparse_ordering.hpp

```diff
@@ -0,0 +1,199 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "common/common.hpp"
+
+#include <algorithm>
+#include <map>
+#include <utility>
+#include <vector>
+
+namespace power_grid_model {
+
+namespace detail {
+inline void remove_element_vector_pair(Idx u, std::vector<std::pair<Idx, Idx>>& dgd) {
+    std::erase_if(dgd, [u](auto const& v) { return v.first == u; });
+}
+
+inline void set_element_vector_pair(Idx u, Idx v, std::vector<std::pair<Idx, Idx>>& dgd) {
+    if (auto it = std::ranges::find_if(dgd, [u](auto const& value) { return value.first == u; }); it != dgd.end()) {
+        it->second = v;
+    }
+}
+
+inline IdxVector adj(Idx const u, std::map<Idx, IdxVector> const& d) {
+    IdxVector l;
+
+    for (const auto& [k, adjacent] : d) {
+        if (k == u) {
+            l.insert(l.end(), adjacent.cbegin(), adjacent.cend());
+        } else if (std::ranges::find(adjacent, u) != adjacent.cend()) {
+            l.push_back(k);
+        }
+    }
+
+    return l;
+}
+
+inline std::vector<std::pair<Idx, std::vector<std::pair<Idx, Idx>>>>
+comp_size_degrees_graph(std::map<Idx, IdxVector> const& d) {
+    std::vector<std::pair<Idx, Idx>> dd;
+    IdxVector v;
+    Idx n = 0;
+
+    for (auto const& [k, adjacent] : d) {
+        if (std::ranges::find(v, k) == v.end()) {
+            ++n;
+            v.push_back(k);
+            dd.emplace_back(k, adj(k, d).size());
+        }
+        for (Idx const e : adjacent) {
+            if (std::ranges::find(v, e) == v.end()) {
+                ++n;
+                v.push_back(e);
+                dd.emplace_back(e, adj(e, d).size());
+            }
+        }
+    }
+
+    std::ranges::sort(dd);
+
+    return {{n, dd}};
+}
+
+inline std::map<Idx, IdxVector> make_clique(IdxVector& l) {
+    std::map<Idx, IdxVector> d;
+
+    for (Idx i = 0; i < static_cast<Idx>(l.size()) - 1; i++) {
+        Idx const idx = i + 1;
+        IdxVector sl(l.size() - idx);
+        std::copy(l.begin() + idx, l.end(), sl.begin());
+        d[l[i]] = std::move(sl);
+    }
+
+    return d;
+}
+
+inline std::vector<std::pair<IdxVector, IdxVector>> check_indistguishable(Idx const u,
+                                                                          std::map<Idx, IdxVector> const& d) {
+    IdxVector rl;
+
+    auto l = adj(u, d);
+    auto lu = l;
+    lu.push_back(u);
+
+    for (auto const& v : l) {
+        auto lv = adj(v, d);
+        lv.push_back(v);
+        std::ranges::sort(lu);
+        std::ranges::sort(lv);
```

Comment:
This `sort` right after appending looks less than ideal

Reply:
i think they have the same complexity, but since the amount of nodes may be quite large, appending before sorting is usually better. see e.g. https://stackoverflow.com/questions/168891/is-it-faster-to-sort-a-list-after-inserting-items-or-adding-them-to-a-sorted-lis .

if we really want to go for it here, using a heap (as also mentioned there) might be even more optimal, but in this case, that's not where the complexity is

Reply:
Sorting after appending each item results in an arithmetic series, i.e., N*(N+1)/2 many in the best case.

Reply:
after investigating, this is actually not that bad, but you're right in that the logic is convoluted. I will investigate to see whether I can make it more readable (and maybe more efficient, as well).

As to why it is not immediately terrible: the vector only contains adjacent edges, and the list of adjacent edges is expected to be small. In addition, the vector is newly created, so `push_back` and then sorting is, in fact, the most efficient way. That said, I don't see any reason why the full list of adjacent edges for every adjacent edge needs to be recreated every single time, but I will look into that.

Reply:
the only trivially straightforward optimization i see is to do the sorting of `lu` outside the loop. all the rest would probably result in more convoluted code

Reply:
Indeed